File: agentic-clj/test/agentic/timers_test.clj
(ns agentic.timers-test
  "Phase 2: portable timers. Logical time, deterministic ordering, durable-via-store, and firing due
   timers back through the stream runtime as turns. Mirrors jagentic-core's timer goldens."
  (:require [clojure.test :refer [deftest is testing]]
            [clojure.string :as str]
            [agentic.core :as core]
            [agentic.event :as ev]
            [agentic.store :as store]
            [agentic.stream :as stream]
            [agentic.timers :as timers]))

(deftest schedule-advance-and-next-deadline
  (testing "due timers come out ascending by :fire-at; advance removes them"
    (let [ts (timers/in-memory-timer-service)]
      (timers/schedule ts "a" 100 (ev/event "c" "u" "a"))
      (timers/schedule ts "b" 50 (ev/event "c" "u" "b"))
      (timers/schedule ts "c" 200 (ev/event "c" "u" "c"))
      (is (= 50 (timers/next-deadline ts)))
      (is (= ["b" "a"] (mapv :id (timers/advance-to ts 150))))
      (is (= 200 (timers/next-deadline ts)))
      (is (= [] (timers/advance-to ts 150)) "already-fired timers do not re-fire"))))

(deftest equal-deadlines-keep-schedule-order
  (testing "equal :fire-at preserves insertion order as the stable tie-break"
    (let [ts (timers/in-memory-timer-service)]
      (timers/schedule ts "x" 100 (ev/event "c" "u" "x"))
      (timers/schedule ts "y" 100 (ev/event "c" "u" "y"))
      (timers/schedule ts "z" 100 (ev/event "c" "u" "z"))
      (is (= ["x" "y" "z"] (mapv :id (timers/advance-to ts 100)))))))

(deftest cancel-removes-pending
  (testing "cancel returns truthy once, falsey thereafter; cancelled timer never fires"
    (let [ts (timers/in-memory-timer-service)]
      (timers/schedule ts "a" 100 (ev/event "c" "u" "a"))
      (is (timers/cancel ts "a"))
      (is (not (timers/cancel ts "a")))
      (is (= [] (timers/advance-to ts 1000))))))

(deftest durable-survives-restore
  (testing "a scheduled timer survives a fresh service over the same store"
    (let [kv (store/in-memory-keyed-state-store)
          dts (timers/durable-timer-service kv)]
      (timers/schedule dts "escalate" 500 (ev/event "c9" "alice" "what is my balance?"))
      ;; A NEW durable service over the SAME store, then restore.
      (let [restored (timers/restore (timers/durable-timer-service kv))]
        (is (= 500 (timers/next-deadline restored)))
        (let [due (timers/advance-to restored 500)]
          (is (= 1 (count due)))
          (is (= "c9" (get-in (first due) [:payload :conversation-id])))
          (is (= "what is my balance?" (get-in (first due) [:payload :text]))))))))

(deftest fire-due-timers-through-stream
  (testing "due timer payloads fire back through the stream runtime as turns"
    (let [sr (stream/stream-runtime (core/banking-system))
          ts (timers/in-memory-timer-service)]
      (timers/schedule ts "followup" 1000 (ev/event "c1" "alice" "what is my balance?"))
      (is (= [] (stream/fire-due-timers sr ts 999)) "nothing due before the deadline")
      (let [results (stream/fire-due-timers sr ts 1000)]
        (is (= 1 (count results)))
        (is (= "payments" (:path (first results))))
        (is (str/includes? (:reply (first results)) "1234.56"))))))

File: ports/experimental/go/core/timers_test.go
package core

import (
	"reflect"
	"strings"
	"testing"
)

// ids extracts the timer ids in order, for golden comparison.
func ids(timers []Timer) []string {
	out := make([]string, 0, len(timers))
	for _, t := range timers {
		out = append(out, t.ID)
	}
	return out
}

// TestTimerScheduleAdvanceOrder mirrors the Java golden: timers come out ascending by
// FireAt, and a drained deadline does not fire twice.
func TestTimerScheduleAdvanceOrder(t *testing.T) {
	ts := NewInMemoryTimerService()
	ts.Schedule("a", 100, NewEvent("c", "u", "a"))
	ts.Schedule("b", 50, NewEvent("c", "u", "b"))
	ts.Schedule("c", 200, NewEvent("c", "u", "c"))

	if d, ok := ts.NextDeadline(); !ok || d != 50 {
		t.Fatalf("NextDeadline = (%d, %v), want (50, true)", d, ok)
	}

	due := ts.AdvanceTo(150)
	if got := ids(due); !reflect.DeepEqual(got, []string{"b", "a"}) {
		t.Fatalf("AdvanceTo(150) ids = %v, want [b a]", got)
	}

	if d, ok := ts.NextDeadline(); !ok || d != 200 {
		t.Fatalf("NextDeadline after fire = (%d, %v), want (200, true)", d, ok)
	}

	if again := ts.AdvanceTo(150); len(again) != 0 {
		t.Fatalf("AdvanceTo(150) again = %v, want empty", ids(again))
	}
}

// TestTimerEqualDeadlinesKeepScheduleOrder asserts equal FireAt values are a stable
// tie-break on schedule order.
func TestTimerEqualDeadlinesKeepScheduleOrder(t *testing.T) {
	ts := NewInMemoryTimerService()
	ts.Schedule("x", 100, NewEvent("c", "u", "x"))
	ts.Schedule("y", 100, NewEvent("c", "u", "y"))
	ts.Schedule("z", 100, NewEvent("c", "u", "z"))

	due := ts.AdvanceTo(100)
	if got := ids(due); !reflect.DeepEqual(got, []string{"x", "y", "z"}) {
		t.Fatalf("equal-deadline order = %v, want [x y z]", got)
	}
}

// TestTimerCancel asserts cancel removes a pending timer, is idempotent on the boolean,
// and prevents the timer from firing.
func TestTimerCancel(t *testing.T) {
	ts := NewInMemoryTimerService()
	ts.Schedule("a", 100, NewEvent("c", "u", "a"))

	if !ts.Cancel("a") {
		t.Fatal("Cancel(a) first time = false, want true")
	}
	if ts.Cancel("a") {
		t.Fatal("Cancel(a) second time = true, want false")
	}
	if due := ts.AdvanceTo(1000); len(due) != 0 {
		t.Fatalf("AdvanceTo(1000) after cancel = %v, want empty", ids(due))
	}
}

// TestTimerReplaceTakesNewScheduleOrder asserts re-scheduling an id replaces it and the
// replacement takes the new schedule order for tie-breaking.
func TestTimerReplaceTakesNewScheduleOrder(t *testing.T) {
	ts := NewInMemoryTimerService()
	ts.Schedule("a", 100, NewEvent("c", "u", "a"))
	ts.Schedule("b", 100, NewEvent("c", "u", "b"))
	// Replace "a" — now scheduled after "b", same deadline.
	ts.Schedule("a", 100, NewEvent("c", "u", "a2"))

	due := ts.AdvanceTo(100)
	if got := ids(due); !reflect.DeepEqual(got, []string{"b", "a"}) {
		t.Fatalf("replaced order = %v, want [b a]", got)
	}
	if len(due) != 2 {
		t.Fatalf("want 2 timers (no duplicate), got %d", len(due))
	}
	if due[1].Payload.Text != "a2" {
		t.Fatalf("replaced payload text = %q, want a2", due[1].Payload.Text)
	}
}

// TestDurableTimerSurvivesRestore asserts a scheduled timer round-trips through a
// KeyedStateStore and fires after a fresh service restores from the same store.
func TestDurableTimerSurvivesRestore(t *testing.T) {
	store := NewInMemoryKeyedStateStore()
	d1 := NewDurableTimerService(store)
	d1.Schedule("escalate", 500, NewEvent("c9", "u9", "what is my balance?"))

	// New service over the SAME store; Restore.
	d2 := NewDurableTimerService(store)
	d2.Restore()

	if dl, ok := d2.NextDeadline(); !ok || dl != 500 {
		t.Fatalf("restored NextDeadline = (%d, %v), want (500, true)", dl, ok)
	}

	due := d2.AdvanceTo(500)
	if len(due) != 1 {
		t.Fatalf("restored AdvanceTo(500) = %d timers, want 1", len(due))
	}
	if due[0].Payload.ConversationID != "c9" {
		t.Fatalf("restored payload ConversationID = %q, want c9", due[0].Payload.ConversationID)
	}
	if due[0].Payload.Text != "what is my balance?" {
		t.Fatalf("restored payload Text = %q, want %q", due[0].Payload.Text, "what is my balance?")
	}
}

// TestFireDueTimersThroughStream asserts a due timer re-enters the StreamRuntime as a
// turn: nothing fires before the deadline, and at the deadline the payload routes through
// the banking graph (payments path, balance reply).
func TestFireDueTimersThroughStream(t *testing.T) {
	timers := NewInMemoryTimerService()
	timers.Schedule("sla", 1000, NewEvent("c1", "alice", "what is my balance?"))

	stream := NewStreamRuntime(NewBankingRuntime())

	if early := stream.FireDueTimers(timers, 999); len(early) != 0 {
		t.Fatalf("FireDueTimers(999) = %d results, want 0", len(early))
	}

	results := stream.FireDueTimers(timers, 1000)
	if len(results) != 1 {
		t.Fatalf("FireDueTimers(1000) = %d results, want 1", len(results))
	}
	if results[0].Path != "payments" {
		t.Fatalf("fired timer path = %q, want payments", results[0].Path)
	}
	if !strings.Contains(results[0].Reply, "1234.56") {
		t.Fatalf("fired timer reply missing balance: %q", results[0].Reply)
	}
}
